fix: Use cross firmware friendly function for getting device version

iperzic opened the pull request:
The originally used [`GetOSVersion`](https://developer.roku.com/en-ca/docs/references/brightscript/interfaces/ifdeviceinfo.md#getosversion-as-object) was added only in 9.2 firmware version

iperzic committed:
fix: handle roku firmware lower than 9.2

MilosRasic commented on components/BugsnagTask.brs:
```diff
@@ -243,7 +243,11 @@ function createDevicePayload()
+	if FindMemberFunction(deviceInfo, "GetOSVersion") <> invalid
+		device["firmwareVersion"] = deviceInfo.GetOSVersion()
+	else 
+		device["firmwareVersion"] = deviceInfo.GetVersion()
```
The docs don't look convincing that this will be the same thing as OS version. Have you been able to check it on a device?

If it's not the OS version, maybe we can just report "unknown" or "pre-9.2" when `GetOSversion()` is not available?

iperzic replied:
You are right, these are not the same. I could do some string manipulation to end up with a similar structure, although it would be missing a revision number.

Now I have realized, `GetOSVersion` returns an object while `GetVersion` returns a string. So I will fix this so that object is set either way for firmwareVersion, but will put revision to `n/a` or something similar. What do you think?

MilosRasic replied:
Sounds good :+1:

iperzic replied:
Refactored in 525284ff ™️ 

Now, either way, an object is returned. Only difference is that for firmware lower than 9.2, revision number is `n/a`.

iperzic committed:
refactor: set object as firmwareVersion, on older devicess

darktasevski commented on components/BugsnagTask.brs:
```diff
@@ -243,7 +243,17 @@ function createDevicePayload()
+		version = deviceInfo.GetVersion()
+		device["firmwareVersion"] = {
+			major: Val(version.mid(2, 1))
+			minor: Val(version.mid(4, 2))
```
The only problem I see here is that we're resolving `40` here as the minor version where it should be `4`. Not sure if adding additional logic to resolve the correct minor version is worth it, thoughts @MilosRasic? This is going to be correct once the devices start upgrading to FW version 10, but with this code we would report a minor version as 40.

MilosRasic replied:
Is this just constraining it to one character? In that case, no additional logical, just do `minor: Val(version.mid(4, 1))` instead and problem solved :)

If not, then what's the logic? I'm not sure I understand tbh. Maybe add a comment here with sample string returned by `deviceInfo.GetVersion()` so that someone reading this in the future can have easier time understanding what we're doing here.

darktasevski replied:
Here's an example of what I'm talking about:

![Screen Shot 2021-03-15 at 5 28 42 PM](https://user-images.githubusercontent.com/25148427/111187049-e99db880-85b3-11eb-92eb-74918257053f.png)

`version` is just regular `GetOSVersion()` which we were using before, concatenated into string.

`_version` is return value of `GetVersion()`, and `__version` is string built with:

```brs
_version = deviceInfo.GetVersion()
__version = _version.mid(2, 1) + " " + _version.mid(4, 2) + " " + _version.mid(8, 4)
```

iperzic replied:
@Puritanic are you sure that `version` is same as `GetOSVersion`? I see no mention of this in the [documentation](https://developer.roku.com/en-ca/docs/references/brightscript/interfaces/ifdeviceinfo.md#getversion-as-string), especially since they mention that `GetVersion` returns **only** major, minor and build number. No mention of the revision number. 

@MilosRasic on the same link, you can find their example that minor can be `08`, so I'd say we will need 2 characters. But yes, some comments here would be nice 🤔

darktasevski replied:
Sorry, I haven't left code impl for the `version` variable, it's built like this:
```brs
osVersion = deviceInfo.GetOSVersion()
version = osVersion.major + "." + osVersion.minor + "." + osVersion.revision + "." + osVersion.build
```

Also, the example in their documentation is not a valid semver, `08` should be just `8`. Maybe we cc. 

@iperzic What can be done here is to use the second number in `40` as a revision number, and that would be actually a correct representation of the device version.

iperzic replied:
But, would it be correct? I know this is not valid semver, but given their example `034.08E01185A` it translates to major 4, minor 08 and build 1185. Maybe I'm following this blindly, or misunderstanding something.

For versions higher than 9.1 this will not be an issue since we'd be using `GetOSVersion` which returns everything correctly.

All this aside, how would you prefer this to be handled, so I can fix this ASAP?

darktasevski replied:
Well, given that `deviceInfo.GetOSVersion()` formatted result is `9.4.0.4200` and `deviceInfo.GetVersion()` formatted result is `9.40.n/a.4200` I would say that it's safe to assume that `0` in `40` is a revision number if we compare these two results. Thoughts @MilosRasic?

MilosRasic replied:
Yeah, let's just consider it `${majorVersion}.${minorVersion}${patchVersion}`. Even if we get it wrong, there's no way we'll get same version numbers with `deviceInfo.GetOSVersion()` on one device and `deviceInfo.GetVersion()` on another.

If usage in production reveals any problems with this, we can fix them later, when we have better info.

darktasevski replied:
@MilosRasic are you saying that we're good with the current implementation @iperzic proposed?

MilosRasic replied:
No, I think in his example, it should be equivalent to 3.0.8, and you screenshot would be 9.4.0

Since we don't report "patch" version the 8 in 3.0.8 is lost and we'd report
major: 3
minor: 0
build: 1185

sound okayish?

iperzic replied:
I will convert the third digit to patch number 👍

iperzic replied:
Refactored in eca834be

darktasevski committed:
refactor: remove function fields and replace them as AA fields

BREAKING CHANGE: this changes BugsnagTask's API, as we're not using function anymore in order to make sure that all code is being executed on the Task thread

MilosRasic committed:
4.0.0

iperzic committed:
refactor: treat 5th version character as revision

MilosRasic commented:
One more thing before we merge @iperzic 

Please add yourself to contributors in package.json

Commit:
chore: add myself to the contributors list

iperzic commented:
Gladly done ✅
